### PR TITLE
feat(core): improve error message for BrowserOnly; better docs

### LINKED ISSUE #6263: BrowserOnly component API has a counterintuitive usage of props.children
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

This is not exactly a bug, more like an API design flaw:  The `<BrowserOnly>` component expects its `props.children` to be a lambda function instead of ordinary JSX elements.  This is counterintuitive and seems unnecessary.

### Steps to reproduce

1. Try to write some code using the `<BrowserOnly>` component:

   ```ts
   // Try 1: "Uncaught TypeError: children is not a function"
   const MyComponent = (props) => {
     return (
       <BrowserOnly>
         {props.children}
       </BrowserOnly>
     );
   };

   // Try 2: Okay let's try the example from the website docs.
   // But still: "Uncaught TypeError: children is not a function"
   const MyComponent = () => {
     return (
       <BrowserOnly>
         {() => <span>page url = {window.location.href}</span>}
         {props.children}
       </BrowserOnly>
     );
   };

   // Try 3: Let's simplify it down to the bare minimum
   // "Uncaught TypeError: children is not a function"
   const MyComponent = () => {
     return (
       <BrowserOnly>
         Okay seriously what is going on here??
       </BrowserOnly>
     );
   };

   // Try 4: An accidental space character 🥴
   // "Uncaught TypeError: children is not a function"
   const MyComponent = () => {
     return (
       <BrowserOnly> </BrowserOnly>
     );
   };
   ```

2. After about 20 minutes of debugging and scrutinizing the docs and banging my head against the wall, what I realized is that this API expects the JSX child content to evaluate to a ***lambda function that returns elements***, not a regular JSX element.  Any stray space character will cause `props.children()` to fail when `BrowserOnly`  tries to invoke the object as a function.

   This API design does work if used correctly, but it goes against all intuition for how JSX notation is supposed to function.

 

## Suggested redesign

Instead of this:

**Current API design**
```ts
const MyComponent = (props) => {
  return (
    <BrowserOnly fallback={<div>Loading...</div>}>
      {() => {
        const LibComponent = require('some-lib').LibComponent;
        return <LibComponent {...props} />;
      }}
    </BrowserOnly>
  );
};
```

...what if the API was like this:

**Proposed API design**
```ts
const MyComponent = (props) => {
  const renderForBrowser = () => {
     const LibComponent = require('some-lib').LibComponent;
     return <LibComponent {...props} />;
  };
  return (
    <BrowserOnly renderForBrowser={renderForBrowser}>
      <div>Loading...</div>
    </BrowserOnly>
  );
};
```

  

### Self-service

- [X] I'd be willing to fix this bug myself.

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

Fix #6263

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added unit test

### CODE REVIEW

File: website/docs/docusaurus-core.md [hunk 1]
```diff
@@ -228,6 +228,29 @@ const MyComponent = (props) => {
 };
 ```
 
+:::info Why do we use a "render function"?
+
+It's important to realize that the children of `<BrowserOnly>` is not a JSX element, but a function that _returns_ an element. This is a design decision. Consider this code:
```
Comment (slorber) [inline website/docs/docusaurus-core.md:233]: 👍 

Maybe worth explaining too that it's a function that is executed as is and not a React component declared inline (bad practice)? (ie you can't call React hooks in this function)

It's up to you.

**Review by slorber — APPROVED** [review]
Thanks 👍 

IMHO the API part of a doc should be objective and factual, not guide/learning/tips oriented so I'd rather keep such infos very short, but I don't know where else we could put that information for now... We can merge this and do some cleanup later 🤷‍♂️

Just wanted to let you know how docs are currently organized. Worth taking a look at https://documentation.divio.com/

**Comment by Josh-Cena** [comment]
I'm doing lots of doc rewriting in #6296 and yes, I do intend to keep the API as API-only. But for now we document `BrowserOnly` nowhere else so I have to put it here. After #6296, I will probably move this to where we discuss SSR

File: website/docs/docusaurus-core.md (hunk 1 above)
Comment (Josh-Cena) [reply website/docs/docusaurus-core.md:233]: I'll still keep it brief😄 Calling hooks will probably throw "invalid hook call" anyways, and once I find a better place to document this, we'll elaborate on that.

**Pull request merged**